neqo-transport/src/pmtud.rs

```diff
@@ -0,0 +1,180 @@
+    4464,  // IEEE 802.5 (4Mb max)          RFC 1042
+    4352,  // FDDI (Revised)                RFC 1188
+    2048,  // Wideband Network              RFC 907
+    2002,  // IEEE 802.5 (4Mb recommended)  RFC 1042
+    1536,  // Exp. Ethernet Nets            RFC 895
+    1500,  // Ethernet Networks             RFC 894
+    1492,  // IEEE 802.3                    RFC 1042
+    1280,  // IPv6 minimum MTU              RFC 2460
+];
+
+// From https://datatracker.ietf.org/doc/html/rfc8899#section-5.1.2
+const MAX_PROBES: usize = 3;
+
+// const PATH_MTU_V6: usize = 1337;
+// const PATH_MTU_V4: usize = PATH_MTU_V6 + 20;
+
+#[derive(Debug, PartialEq)]
+pub enum Probe {
+    NotNeeded,
+    Needed,
+    Prepared,
+    Sent,
+}
+
+#[derive(Debug)]
+pub struct Pmtud {
+    header_size: usize,
+    mtu: usize,
+    probed_index: usize,
+    probe_count: usize,
+    probe_state: Probe,
+}
+
+pub type PmtudRef = Rc<RefCell<Pmtud>>;
+
+impl Pmtud {
+    #[must_use]
+    pub fn new(remote_ip: IpAddr) -> PmtudRef {
+        Rc::new(RefCell::new(Pmtud {
+            header_size: Self::header_size(remote_ip),
+            mtu: MTU_SIZES[MTU_SIZES.len() - 1],
+            probed_index: MTU_SIZES.len() - 1,
+            probe_count: 0,
+            probe_state: Probe::NotNeeded,
+        }))
+    }
+
+    #[must_use]
+    pub fn plpmtu(&self) -> usize {
+        self.mtu - self.header_size
+    }
+
+    #[must_use]
+    pub fn needs_probe(&self) -> bool {
+        self.probe_state == Probe::Needed
+    }
+
+    #[must_use]
+    pub fn is_probe_prepared(&self) -> bool {
+        self.probe_state == Probe::Prepared
+    }
+
+    pub fn probe_prepared(&mut self) {
+        self.probe_state = Probe::Prepared;
+        qtrace!(
+            "PMTUD probe of size {} prepared",
+            MTU_SIZES[self.probed_index]
+        );
+    }
+
+    pub fn probe_sent(&mut self, stats: &mut Stats) -> bool {
+        self.probe_state = Probe::Sent;
+        self.probe_count += 1;
+        stats.pmtud_tx += 1;
+        qdebug!(
+            "PMTUD probe of size {} sent, count {}",
+            MTU_SIZES[self.probed_index],
+            self.probe_count
+        );
+        true
+    }
+
+    #[must_use]
+    pub fn is_pmtud_probe(&self, p: &SentPacket) -> bool {
+        self.probe_state == Probe::Sent
+            && p.len() == MTU_SIZES[self.probed_index] - self.header_size
+    }
+
+    pub fn on_packets_acked(&mut self, acked_pkts: &[SentPacket], stats: &mut Stats) {
+        if self.probe_state != Probe::Sent
+            || acked_pkts.is_empty()
+            || !acked_pkts.iter().any(|p| self.is_pmtud_probe(p))
+        {
+            return;
+        }
+        stats.pmtud_ack += 1;
+        self.mtu = MTU_SIZES[self.probed_index];
+        qdebug!(
+            "PMTUD probe of size {} succeeded",
+            MTU_SIZES[self.probed_index]
+        );
+        self.start_pmtud();
+    }
+
+    pub fn on_packets_lost(&mut self, lost_packets: &[SentPacket], stats: &mut Stats) {
```

Comment:
This is a useful function, but I think it could be more general.

If, at any time, at least N packets of size greater than a previous size threshold are lost, then we reduce the MTU to that threshold.

That threshold can be lower when probing up than when dropping down, but the basic theory should work.